### ISSUE TITLE
[BUG] rspamd_http lua module does not handle AAAA DNS records

### ISSUE THREAD

**Issue by TheEnbyperor** [rspamd/rspamd#4207]
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Read about bug reporting in general: https://rspamd.com/doc/faq.html#how-to-report-bugs-found-in-rspamd
    * Enabled relevant debugging logs: https://rspamd.com/doc/faq.html#how-to-debug-some-module-in-rspamd 
    * Checked the FAQs about Core files in case of fatal crash: https://rspamd.com/doc/faq.html#how-to-figure-out-why-rspamd-process-crashed
    * Tried ASAN package and obtained the ASAN report (if possible): https://rspamd.com/doc/faq.html#asan-builds 
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=%E2%9C%93&q=is%3Aissue+user%3Arspamd
    * Checked that there is not already an experimental package or master branch

**Describe the bug**
The `rspamd_http` makes no attempt to resolve AAAA records of domains its asked to connect to.

### Steps to Reproduce
In the `static gint lua_http_request (lua_State *L)` function of the `lua_http.c` file when a domain name must be resolved to connect to the following code is run
```c
if (!rspamd_dns_resolver_request (resolver, session, NULL, lua_http_dns_handler, cbd, RDNS_REQUEST_A, cbd->host)) {
```
No attempt to resolve AAAA records is ever made.

Its clear from earlier in the same file in the `static void lua_http_dns_handler (struct rdns_reply *reply, gpointer ud)` function that this was not intended as there is code to handle AAAA responses, but these will never be produced in response to a A only query.
```c
if (reply->entries->type == RDNS_REQUEST_A) {
	cbd->addr = rspamd_inet_address_new (AF_INET,
			&reply->entries->content.a.addr);
} else if (reply->entries->type == RDNS_REQUEST_AAAA) {
	cbd->addr = rspamd_inet_address_new (AF_INET6,
			&reply->entries->content.aaa.addr);
}
```

### Versions
Affects all OSes, bug present in current master branch.

**Comment by vstakhov** [comment]
Well, I have been watching closely for IPv6 progress for more than a decade, as I was utterly surprised by such a huge effort to resolve non-existing problem. And so far, in 2022, IPv6 is still a dead space despite of increased traffic (mostly because of Google/Facebook), as there is no life outside of IPv4 - you just cannot afford to have IPv6 only resource, as you would loose a huge portion of the auditory (including myself as wire providers in the UK do not care about IPv6). So what you offer is to send an extra DNS request to support something that should not exist in the first place (at least in the way it's designed, of course). I do agree that it could be reasonable in some environments but in general Rspamd should just use IPv4 as it works everywhere without any issues. In general, I would say that IPv6 should be a flag that can be turned on but is left turned off by default. Of course, we are talking about email and high performance in the first place.

**Comment by TheEnbyperor** [comment]
I've made this point so many times over my life its getting ludicrous. Even if the general consumer broadband to service edge internet remains IPv4 (which it won't, but you seem adamant against that, so I won't argue further) there is plenty of IPv6 only service to service communication. 

We as a company have one /24 IPv4 allocation, and there is little to no prospect of getting any more from RIPE. All our services communicate with each other over v6 only because they do not need to be accessible to absolutely everyone on the internet, but still require global addressability. This bug was discovered whilst trying to use the vault feature of the DKIM signing. Our vault server has no v4 (it doesn't need it) and as such rspamd was completely unable to communicate without pasting the raw v6 address into the config.

IPv6 has been a standard since 1998 and fully ratified as the way forward for the internet since 2017. It is downright incompetent to suggest nobody should care about v6 whatsoever.

**Comment by kescherCode** [comment]
@vstakhov

> to resolve non-existing problem

There are more internet-connected devices on the internet than IPv4 addresses. This is bad, and must be addressed. This was foreseen in 1998, and has become **last decade.

> So what you offer is to send an extra DNS request to support something that should not exist in the first place (the way it is designed, of course)

See above, and literally every other software supporting IPv6 does just that. "It should not exist in the first place" is a moot point.

> In general, I would say that IPv6 should be a flag that can be turned on but is left turned off by default. Of course, we are talking about email and high performance in the first place.

Dovecot and Postfix support IPv4 and IPv6. This software is usually used in tandem with rspamd. So, no. Support IPv6, end of story.

**Comment by basisbit** [comment]
+1 request for IPv6 support in rspamd. Pls fix bugs - it is not even that hard...

**Comment by Manawyrm** [comment]
We see over 70% of email traffic via IPv6 nowadays (in germany, for mostly private customers), all internal infrastructure only uses IPv6. Only the machines at the "edge" have IPv4 addresses.
While it's possible that IPv6 hasn't really taken off/over in the UK, things are looking very different in other parts of the world.
AAAA lookup support (at least optionally) would be ideal. 
The current setup uses private IPv4 addressing over IPv6 tunnels using Wireguard, which does work, but that's just an ugly hack.

**Comment by vstakhov** [comment]
> We as a company have one /24 IPv4 allocation, and there is little to no prospect of getting any more from RIPE. All our services communicate with each other over v6 only because they do not need to be accessible to absolutely everyone on the internet, but still require global addressability. This bug was discovered whilst trying to use the vault feature of the DKIM signing. Our vault server has no v4 (it doesn't need it) and as such rspamd was completely unable to communicate without pasting the raw v6 address into the config.

Well, I totally agree that there are IPv6 only networks where Rspamd should just use IPv6 in DNS lookups. And the change in the code is trivial: check for some flag and query for IPv6 if it is enabled. What I will not do is to add support dual-stack in happy-eyeballs like approach.

**Comment by vstakhov** [comment]
> We see over 70% of email traffic via IPv6 nowadays (in germany, for mostly private customers), all internal infrastructure only uses IPv6. Only the machines at the "edge" have IPv4 addresses.

Rspamd perfectly supports IPv6 in email traffic and DNS requests. The only question is how to query external services via HTTP/TCP. I hate the idea to add dual-stack support there for many reasons (stability, extra load, error prone, hard to handle in the asynchronous code... etc). The ability to switch between IPv4/IPv6 mode by means of the configuration knob is, on the contrary, a very easy thing to add and maintain.

If that makes you more happy then I will add it. 

But please don't start a holy war about dual-stack here. If you really want to see it, please submit pull requests for the review. I personally will not invest any of my time into dual-stack support as I consider this stuff a big technical mistake (not IPv6 per se, but the way how it interacts with IPv4).

**Comment by vstakhov** [comment]
> Could be trivially done by connecting each address and retrying with the next on failure.

Huh, trivial? Then submit a PR.

**Comment by vstakhov** [comment]
> What part exactly do you plan not to implement?

[Happy eyeballs](https://datatracker.ietf.org/doc/html/rfc6555)

There are whole bunch of things that got broken in the async world. Let's imagine that you have 3 stages:

* DNS resolution
* TCP connection (+ probably tls)
* Data transfer

With just IPv4 and IPv6 this is a DAG: you just move throughout the stages. With happy eyeballs you need to maintain two individual requests and stop one at the data transfer stage. So it will look like this:

```
IPv4 DNS            IPv6 DNS
 |                          |
IPv4 connect      IPv6 connect
 |                        /
Sync point.....
|                          |
IPv4 close         IPv6 data transfer
                           |
                          IPv6 close
```

In async world this has proven to be much harder (more refcounts etc) and more error prone: for example, your DNS can return IPv6 addresses but you might have only link-local connection for IPv6 (real stories from the past). Or, even worse, you can have slow/huge latency ipv6 and just hang slowing down the overall connection progress.

These are real users' stories, and, indeed, I care more about the vast majority of Rspamd users than about some `frustrated` users who are not lazy enough to write some harsh words to the author of the free and OpenSource project...

So my suggestion is simple: I can implement a configuration switch to work in ipv4 or ipv6 modes only. If you still need support for the dual-stack resolving, then please invest your time more wisely than trying to insult me: write this code, write tests, submit PR, and I will review it.

**Comment by vstakhov** [comment]
> Thank you for your work! I think rspamd is a great project and there should not be conflicts and a fork because of such a small issue.
> 
> I think the problem is that people who are using IPv6 for a very long time in production keep hitting walls on a daily basis where things don't work as expected, although the protocol should really be supported everywhere by now. I hope you can understand the frustration there that sometimes leads to harsh words.

Thank you, I also think there could be some sane consensus on that topic. 

As a maintainer I have to find a way to support different people. Some are frustrated about lack of IPv6 support, some are frustrated that their setup got broken because of that IPv6 support... We are also not talking about the whole Rspamd which supports IPv6 from the beginning but about two Lua modules (specifically, lua_http and lua_tcp) that are usually intended to work with some internal services and they are designed to be efficient and simple. But it has a drawback in terms of flexibilty indeed. And I'm keen to resolve this problem, but not in a way that may damage other users. There are many options to do so: a global ipv4/ipv6 switch, an extra parameter to `lua_http.request`... However, the comments like 'do this, end of story' just demotivates me from doing that for apparent reasons.

**Comment by mutax** [comment]
Thats what I meant with overhead - when you implement HE fully you would indeed need to add a lot of state and do requests in parallel. 

My comment regarding 'that should be trivial' was meant with regards of just "adding v6 support" and allow a fallback.
You could already receive multiple IP-adresses in a A lookup and even if the resolver does round robin, one should try all of these adresses, in case one of them is down, right? So it should be relatively easy to add the result of the AAAA lookup to this list of returned adresses and in case there is a connection failure use the next one. 
This way you would be supporting v4 and v6, allow fallback in case one the protocols is not working, etc.

I hope that clarifies what I meant. I totally agree that implementing HE completely is a waste of time here, but supporting Dualstack is more than a nice to have in my opinion and will on the long run be easier to maintain.

I am not sure if a switch that enables v4 OR v6 is actually that helpful and will probably at some point just result in tickets by confused users that don't understand why rspamd behaves differently in this subcomponent and differently in the rest.

**Comment by vstakhov** [comment]
> You could already receive multiple IPv4-addresses in a A lookup and even if the resolver does round robin, one should try all of these addresses, in case one of them is down, right? So it should be relatively easy to add the result of the AAAA lookup to this list of returned addresses and in case there is a connection failure use the next one.
> This way you would be supporting v4 and v6, allow fallback in case one the protocols is not working, etc.

There is one problem with this approach: it requires an extra DNS request all the time you resolve the service name. Unfortunately, it has caused severe issues for some large Rspamd users in the past. 

Curiously, when I was writing `librdns` in the past, I was under impression that DNS servers really support RFC 1035 in terms of multiple queries per request. But in fact, my library is the only client library I'm aware of that support multiple queries per request and can query for both A and AAAA records in a single packet. Unfortunately, it is not supported by any DNS server software apart from (surprisingly) dnsmasq.

So if there is a feasible solution without two DNS requests per HTTP request, I would be interested to consider it.

**Comment by mutax** [comment]
> There is one problem with this approach: it requires an extra DNS request all the time you resolve the service name. Unfortunately, it has caused severe issues for some large Rspamd users in the past.

I see the issue here.

If I understand the code correctly, you are resolving each time there is a request to be made using the configured resolver, which means not caching the responses locally but this way leveraging the round robin behavior of the upstream resolver. (question for clarification, not questioning the design decision here)

> So if there is a feasible solution without two DNS requests per HTTP request, I would be interested to consider it.

I fear to do A and AAAA one would need to do two independent requests and the only way of reducing these would be to have a LRU cache or similar in the code which indeed adds more complexity that would normally be handled by the stub resolver on the system.

One could introduce a parameter to the lua call that configures the behavior, e.g. do both queries or only A or only AAAA, so people that fear the extra load on the DNS can disable it. I however see that enabling dual stack lookups by default - which might make sense globally for new deployments, this could introduce a breaking change for people that don't want that extra load on the DNS. Caching the negative answers could help in that case?

Yet another way would be to make the A lookup, try to connect, and if that fails do the AAAA lookup, with a switch to turn that behaviour around for IPv6-only environments. Sounds more of a hack but could reduce queries when needed.

**Comment by vstakhov** [comment]
> I do see that there already is an LRU cache for the answers

No, there is no such a thing so far - there is a cache merely for DNSFAIL responses that was designed to resolve one particular issue for a large Rspamd user.

But actually after thinking about this problem, I've found that there is a completely easy and straightforward way to do all this stuff transparently and efficiently.

Instead of modifying the existing C code (which is meh...) it is possible to create a Lua module (like `lua_redis` for `rspamd_redis`) that will process url's and maintain everything in [`upstreams`](https://rspamd.com/doc/configuration/upstream.html). This module provides transparent resolving, lazy re-resolving, re-resolving on errors, addresses sorting and so on. In fact, we just need to process an url, extract hostname and pass it to the upstreams module. After that, each request will use IP address instead of the hostname, as upstreams will perform names resolution in background. This will cover all existing issues with IPv6/IPv4 confusing without touching C parts. Furthermore, it will actually improve the performance as instead of sending a DNS request on each request, Lua will just use a cached response that is periodically refreshed. It will also give control over IPv6/IPv4 errors, as ustreams were designed to deal with this issue at some extent. It will also be possible to implement retransmits/retries logic with backpressure if there is any need of such a logic.
